applications/FluidDynamicsApplication/custom_elements/surface_smoothing_element.cpp

```diff
@@ -0,0 +1,670 @@
+                                    this->pGetGeometry()->pGetPoint(mNode0ID[i_ne]),
+                                    this->pGetGeometry()->pGetPoint(mNode1ID[i_ne]),
+                                    this->pGetGeometry()->pGetPoint(mNode2ID[i_ne]));
+
+            unsigned int contact_node = 0;
+            for (unsigned int i=0; i < num_face_nodes; ++i){
+                if ( outer_face[i].GetValue(IS_STRUCTURE) == 1.0 ){
+                    contact_node++;
+                }
+            }
+
+            if (contact_node == num_face_nodes){
+                auto IntegrationMethod = GeometryData::GI_GAUSS_1;
+                auto face_gauss_pts = outer_face.IntegrationPoints(IntegrationMethod);
+                const unsigned int num_int_pts = face_gauss_pts.size();
+
+                std::vector < GeometryType::CoordinatesArrayType > face_gauss_pts_gl_coords, face_gauss_pts_loc_coords;
+                face_gauss_pts_gl_coords.clear();
+                face_gauss_pts_loc_coords.clear();
+                face_gauss_pts_gl_coords.reserve(num_int_pts);
+                face_gauss_pts_loc_coords.reserve(num_int_pts);
+
+                VectorType face_jacobians;
+                outer_face.DeterminantOfJacobian(face_jacobians, IntegrationMethod);
+
+                // Get the original geometry shape function and gradients values over the intersection
+                for (unsigned int i_gauss = 0; i_gauss < num_int_pts; ++i_gauss) {
+                    // Store the Gauss points weights
+                    const double face_weight = face_jacobians(i_gauss) * face_gauss_pts[i_gauss].Weight();
+
+                    // Compute the global coordinates of the face Gauss pt.
+                    GeometryType::CoordinatesArrayType global_coords = ZeroVector(num_dim);
+                    global_coords = outer_face.GlobalCoordinates(global_coords, face_gauss_pts[i_gauss].Coordinates());
+
+                    // Compute the parent geometry local coordinates of the Gauss pt.
+                    GeometryType::CoordinatesArrayType loc_coords = ZeroVector(num_dim);
+                    loc_coords = GetGeometry().PointLocalCoordinates(loc_coords, global_coords);
+
+                    // Compute shape function values
+                    // Obtain the parent subgeometry shape function values
+                    Kratos::Vector face_shape_func;
+                    face_shape_func = GetGeometry().ShapeFunctionsValues(face_shape_func, loc_coords);
+
+                    double temp_value = 0.0;
+
+                    for (unsigned int i = 0; i < num_nodes; i++){
+
+                        VectorType solid_normal = ZeroVector(num_dim);
+                        if (GetGeometry()[i].GetValue(IS_STRUCTURE) == 1.0)
+                        {
+                            solid_normal = GetGeometry()[i].FastGetSolutionStepValue(NORMAL);
+                            const double norm = Kratos::norm_2(solid_normal);
```

Comment:
Maybe we can check that this is not close to zero in debug mode.
```
#ifdef KRATOS_DEBUG
    KRATOS_WARNING_IF("DistanceSmoothingElement", norm < 1.0e-12) << "WARNING: Normal close to zero" <<std::endl;
#endif
```

Reply:
Done.